crates/polars-utils/src/cache.rs

```diff
@@ -29,3 +38,197 @@ where
+                return slot.value.assume_init_mut();
+            }
+
+            let key = key.to_owned();
+            let val = f(&key);
+            self.raw_insert(h, key, val)
+        }
+    }
+
+    unsafe fn raw_get<Q: Eq + ?Sized>(&self, h: HashResult, key: &Q) -> Option<usize>
+    where
+        K: Borrow<Q>,
+    {
+        unsafe {
+            // SAFETY: we assume h is a HashResult from self.hash with valid indices
+            // and we check slot.last_access != 0 before assuming the slot is initialized.
+            let slot = self.slots.get_unchecked(h.i1);
+            if slot.last_access.get() != 0
+                && slot.hash_tag == h.tag
+                && slot.key.assume_init_ref().borrow() == key
+            {
+                slot.last_access.set(self.new_access_ctr());
+                return Some(h.i1);
+            }
+
+            let slot = self.slots.get_unchecked(h.i2);
+            if slot.last_access.get() != 0
+                && slot.hash_tag == h.tag
+                && slot.key.assume_init_ref().borrow() == key
+            {
+                slot.last_access.set(self.new_access_ctr());
+                return Some(h.i2);
+            }
+        }
+
+        None
+    }
+
+    unsafe fn raw_insert(&mut self, h: HashResult, key: K, value: V) -> &mut V {
+        let last_access = self.new_access_ctr();
+        unsafe {
+            // SAFETY: i1 and i2 are valid indices and older_idx returns one of them.
+            let idx = self.older_idx(h.i1, h.i2);
+            let slot = self.slots.get_unchecked_mut(idx);
+
+            // Drop impl takes care of dropping old value, if occupied.
+            *slot = CacheSlot {
+                last_access: Cell::new(last_access),
+                hash_tag: h.tag,
+                key: MaybeUninit::new(key),
+                value: MaybeUninit::new(value),
+            };
+            slot.value.assume_init_mut()
+        }
+    }
+
+    /// Returns the older index based on access time, where unoccupied slots
+    /// are considered infinitely old.
+    unsafe fn older_idx(&mut self, i1: usize, i2: usize) -> usize {
+        let age1 = self.slots.get_unchecked(i1).last_access.get();
+        let age2 = self.slots.get_unchecked(i2).last_access.get();
+        match (age1, age2) {
+            (0, _) => i1,
+            (_, 0) => i2,
+            // This takes into account the wrap-around of our access_ctr.
+            _ if age1.wrapping_sub(age2) >= (1 << 31) => i1,
+            _ => i2,
+        }
+    }
+
+    fn new_access_ctr(&self) -> u32 {
+        // This keeps the access_ctr always odd, so we don't hit access_ctr == 0,
+        // which would leak values.
+        self.access_ctr.replace(self.access_ctr.get() + 2)
+    }
+
+    /// Computes the hash tag and two slot indexes for a given key.
+    fn hash<Q: Hash + ?Sized>(&self, key: &Q) -> HashResult {
+        // These are just two randomly chosen odd 64-bit numbers.
```

Comment:
Out of curiousity. Does it matter what numbers we sample randomly for the hash strength?

Reply:
It doesn't really matter, as long as they're odd. You could initialize the random numbers using `RandomState` once if you really wanted to ensure no one could potentially exploit the relationship between the two.